diff --git a/src/Makefile.in b/src/Makefile.in
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -55,7 +55,7 @@ am__installdirs = "$(DESTDIR)$(bindir)"
 PROGRAMS = $(bin_PROGRAMS)
 am_nano_OBJECTS = browser.$(OBJEXT) chars.$(OBJEXT) color.$(OBJEXT) \
 	cut.$(OBJEXT) files.$(OBJEXT) global.$(OBJEXT) help.$(OBJEXT) \
-	move.$(OBJEXT) nano.$(OBJEXT) prompt.$(OBJEXT) \
+	move.$(OBJEXT) nano.$(OBJEXT) nano_pepper.$(OBJEXT) prompt.$(OBJEXT) \
 	rcfile.$(OBJEXT) search.$(OBJEXT) text.$(OBJEXT) \
 	utils.$(OBJEXT) winio.$(OBJEXT)
 nano_OBJECTS = $(am_nano_OBJECTS)
@@ -203,6 +203,7 @@ nano_SOURCES = browser.c \
 		help.c \
 		move.c \
 		nano.c \
+		nano_pepper.c \
 		nano.h \
 		prompt.c \
 		proto.h \
@@ -286,7 +287,7 @@ clean-binPROGRAMS:
 	-test -z "$(bin_PROGRAMS)" || rm -f $(bin_PROGRAMS)
 nano$(EXEEXT): $(nano_OBJECTS) $(nano_DEPENDENCIES) 
 	@rm -f nano$(EXEEXT)
-	$(LINK) $(nano_OBJECTS) $(nano_LDADD) $(LIBS)
+	$(LINK) $(nano_OBJECTS) $(nano_LDADD) $(LIBS) $(EXTRA_LIBS)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
diff --git a/src/nano.c b/src/nano.c
--- a/src/nano.c
+++ b/src/nano.c
@@ -1279,6 +1279,14 @@ RETSIGTYPE handle_sigwinch(int signal)
     LINES = win.ws_row;
 #endif
 
+#ifndef __native_client__
+    // NaCl's signal hanlder currently runs on the main PPAPI thread
+    // but nano expects it to run on the nano thread and be able to
+    // redraw the script from within the handler, which is currently
+    // no supported.
+    return;
+#endif
+
     /* If we've partitioned the filestruct, unpartition it now. */
     if (filepart != NULL)
 	unpartition_filestruct(&filepart);
@@ -2013,7 +2021,11 @@ void do_output(char *output, size_t output_len, bool allow_cntrls)
 	update_line(openfile->current, openfile->current_x);
 }
 
+#if defined(__native_client__)
+int nacl_nano_main(int argc, char **argv)
+#else
 int main(int argc, char **argv)
+#endif
 {
     int optchr;
     ssize_t startline = 1;
