diff --git a/libbb/appletlib.c b/libbb/appletlib.c
--- a/libbb/appletlib.c
+++ b/libbb/appletlib.c
@@ -772,7 +772,10 @@ void FAST_FUNC run_applet_and_exit(const char *name, char **argv)
 
 
 
-#if ENABLE_BUILD_LIBBUSYBOX
+#if defined(__native_client__)
+#include "ppapi_simple/ps_main.h"
+int nacl_ppapi_main(int argc UNUSED_PARAM, char **argv)
+#elif ENABLE_BUILD_LIBBUSYBOX
 int lbb_main(char **argv)
 #else
 int main(int argc UNUSED_PARAM, char **argv)
@@ -828,3 +831,7 @@ int main(int argc UNUSED_PARAM, char **argv)
 	xfunc_die();
 #endif
 }
+
+#if defined(__native_client__)
+PPAPI_SIMPLE_REGISTER_MAIN(nacl_ppapi_main)
+#endif
