diff --git a/test/autogen.sh b/test/autogen.sh
--- a/test/autogen.sh
+++ b/test/autogen.sh
@@ -3,7 +3,7 @@
 # Regenerate configuration files
 cp acinclude.m4 aclocal.m4
 found=false
-for autoconf in autoconf autoconf259 autoconf-2.59
+for autoconf in autoconf2.64 autoconf autoconf259 autoconf-2.59
 do if which $autoconf >/dev/null 2>&1; then $autoconf && found=true; break; fi
 done
 if test x$found = xfalse; then
diff --git a/test/configure.in b/test/configure.in
--- a/test/configure.in
+++ b/test/configure.in
@@ -67,7 +67,7 @@ case "$host" in
         ;;
     *-*-nacl*)
         case "$CC" in
-            *pnacl*)
+            */pnacl-*)
                 EXE=".pexe"
                 ;;
             *)
@@ -86,6 +86,8 @@ case "$host" in
         SYS_GL_LIBS="-lGL"
         ;;
 esac
+
+echo "----------------> Using EXE=$EXE"
 AC_SUBST(EXE)
 AC_SUBST(MATHLIB)
 AC_SUBST(ISMACOSX)
