diff --git a/Makefile.gnu b/Makefile.gnu
--- a/Makefile.gnu
+++ b/Makefile.gnu
@@ -13,6 +13,8 @@ DOS2UNIX = dos2unix
 
 LIBRARIES = -lstdc++
 
+ENABLE_SHARED ?= 1
+
 MODULES = $(SRCS:.c=.o)
 MODULES := $(MODULES:.cpp=.o)
 CFLAGS ?= -O3 -fPIC -fexceptions -fvisibility=hidden
@@ -49,13 +51,19 @@ all: dist
 dist: FreeImage
 	mkdir -p Dist
 	cp *.a Dist/
+ifeq ($(ENABLE_SHARED),1)
 	cp *.so Dist/
+endif
 	cp Source/FreeImage.h Dist/
 
 dos2unix:
 	@$(DOS2UNIX) $(SRCS) $(INCLS)
 
+ifeq ($(ENABLE_SHARED),1)
 FreeImage: $(STATICLIB) $(SHAREDLIB)
+else
+FreeImage: $(STATICLIB)
+endif
 
 .c.o:
 	$(CC) $(CFLAGS) -c $< -o $@
@@ -71,11 +79,13 @@ $(SHAREDLIB): $(MODULES)
 
 install:
 	install -d $(INCDIR) $(INSTALLDIR)
-	install -m 644 -o root -g root $(HEADER) $(INCDIR)
-	install -m 644 -o root -g root $(STATICLIB) $(INSTALLDIR)
-	install -m 755 -o root -g root $(SHAREDLIB) $(INSTALLDIR)
+	install -m 644 $(HEADER) $(INCDIR)
+	install -m 644 $(STATICLIB) $(INSTALLDIR)
+ifeq ($(ENABLE_SHARED),1)
+	install -m 755 $(SHAREDLIB) $(INSTALLDIR)
 	ln -sf $(SHAREDLIB) $(INSTALLDIR)/$(VERLIBNAME)
 	ln -sf $(VERLIBNAME) $(INSTALLDIR)/$(LIBNAME)	
+endif
 #	ldconfig
 
 clean:
