diff --git a/Makedefs.in b/Makedefs.in
--- a/Makedefs.in
+++ b/Makedefs.in
@@ -40,14 +40,14 @@ SHELL		=	/bin/sh
 # Installation programs...
 #
 
-INSTALL_BIN	=	$(LIBTOOL) $(INSTALL) -c -m 555 @INSTALL_STRIP@
+INSTALL_BIN	=	$(LIBTOOL) $(INSTALL) -c -m 755 @INSTALL_STRIP@
 INSTALL_COMPDATA =	$(INSTALL) -c -m 444 @INSTALL_GZIP@
 INSTALL_CONFIG	=	$(INSTALL) -c -m @CUPS_CONFIG_FILE_PERM@
 INSTALL_DATA	=	$(INSTALL) -c -m 444
 INSTALL_DIR	=	$(INSTALL) -d
-INSTALL_LIB	=	$(LIBTOOL) $(INSTALL) -c -m 555 @INSTALL_STRIP@
+INSTALL_LIB	=	$(LIBTOOL) $(INSTALL) -c -m 755 @INSTALL_STRIP@
 INSTALL_MAN	=	$(INSTALL) -c -m 444
-INSTALL_SCRIPT	=	$(INSTALL) -c -m 555
+INSTALL_SCRIPT	=	$(INSTALL) -c -m 755
 
 #
 # Default user, group, and system groups for the scheduler...
diff --git a/Makefile b/Makefile
--- a/Makefile
+++ b/Makefile
@@ -20,7 +20,7 @@ include Makedefs
 # Directories to make...
 #
 
-DIRS	=	cups test $(BUILDDIRS)
+DIRS	=	cups $(BUILDDIRS)
 
 
 #
diff --git a/configure b/configure
--- a/configure
+++ b/configure
@@ -5928,6 +5928,10 @@ case "$COMPONENTS" in
 		BUILDDIRS="data locale"
 		;;
 
+  core-and-image)
+    BUILDDIRS="data filter locale"
+    ;;
+
 	*)
 		as_fn_error $? "Bad build component \"$COMPONENT\" specified!" "$LINENO" 5
 		;;
diff --git a/cups/backchannel.c b/cups/backchannel.c
--- a/cups/backchannel.c
+++ b/cups/backchannel.c
@@ -28,6 +28,7 @@
 #  include <sys/time.h>
 #endif /* WIN32 */
 
+#include <sys/select.h>
 
 /*
  * Local functions...
