diff --git a/pixman/pixman-mmx.c b/pixman/pixman-mmx.c
--- a/pixman/pixman-mmx.c
+++ b/pixman/pixman-mmx.c
@@ -97,7 +97,7 @@ _mm_shuffle_pi16 (__m64 __A, int8_t const __N)
 
     asm ("pshufw %2, %1, %0\n\t"
 	: "=y" (ret)
-	: "y" (__A), "K" (__N)
+	: "y" (__A), "X" (__N)
     );
 
     return ret;
diff --git a/pixman/pixman-x86.c b/pixman/pixman-x86.c
--- a/pixman/pixman-x86.c
+++ b/pixman/pixman-x86.c
@@ -80,7 +80,7 @@ detect_cpu_features (void)
 static pixman_bool_t
 have_cpuid (void)
 {
-#if _PIXMAN_X86_64 || defined (_MSC_VER)
+#if _PIXMAN_X86_64 || defined (_MSC_VER) || defined (__native_client__)
 
     return TRUE;
 
