diff --git a/Makefile.in b/Makefile.in
--- a/Makefile.in
+++ b/Makefile.in
@@ -54,7 +54,7 @@ OBJ = \
 all: less$(EXEEXT) lesskey$(EXEEXT) lessecho$(EXEEXT)
 
 less$(EXEEXT): ${OBJ}
-	${CC} ${LDFLAGS} -o $@ ${OBJ} ${LIBS}
+	${CC} ${LDFLAGS} -o $@ ${OBJ} ${NACL_CLI_MAIN_LIB} ${LIBS}
 
 lesskey$(EXEEXT): lesskey.${O} version.${O}
 	${CC} ${LDFLAGS} -o $@ lesskey.${O} version.${O}
diff --git a/main.c b/main.c
--- a/main.c
+++ b/main.c
@@ -64,7 +64,12 @@ extern int	pr_type;
  * Entry point.
  */
 int
-main(argc, argv)
+# if defined(__native_client__)
+nacl_main
+# else
+main
+# endif
+(argc, argv)
 	int argc;
 	char *argv[];
 {
diff --git a/screen.c b/screen.c
--- a/screen.c
+++ b/screen.c
@@ -282,7 +282,8 @@ raw_mode(on)
 			save_term = s;
 			saved_term = 1;
 		}
-#if HAVE_OSPEED
+/*cfgetospeed is not linkable , shouldn't define HAVE_OSPEED */
+#if HAVE_OSPEED &&  !defined(__native_client__)
 		switch (cfgetospeed(&s))
 		{
 #ifdef B0
