diff --git a/Demos/HelloWorld/Makefile.am b/Demos/HelloWorld/Makefile.am
new file mode 100644
--- /dev/null
+++ b/Demos/HelloWorld/Makefile.am
@@ -0,0 +1,5 @@
+noinst_PROGRAMS=AppHelloWorld
+
+AppHelloWorld_SOURCES=HelloWorld.cpp
+AppHelloWorld_CXXFLAGS=-I@top_srcdir@/src $(CXXFLAGS)
+AppHelloWorld_LDADD=-L../../src -lBulletDynamics -lBulletCollision -lLinearMath
diff --git a/Demos/Makefile.am b/Demos/Makefile.am
--- a/Demos/Makefile.am
+++ b/Demos/Makefile.am
@@ -1,5 +1,7 @@
+SUBDIRS=HelloWorld
+if CONDITIONAL_BUILD_DEMOS
+SUBDIRS+=OpenGL BasicDemo TerrainDemo VehicleDemo CcdPhysicsDemo SoftDemo AllBulletDemos
+endif
 if CONDITIONAL_BUILD_MULTITHREADED
-SUBDIRS=OpenGL BasicDemo TerrainDemo VehicleDemo CcdPhysicsDemo MultiThreadedDemo SoftDemo AllBulletDemos
-else
-SUBDIRS=OpenGL BasicDemo TerrainDemo VehicleDemo CcdPhysicsDemo SoftDemo AllBulletDemos
+SUBDIRS+=MultiThreadedDemo
 endif
diff --git a/Makefile.am b/Makefile.am
--- a/Makefile.am
+++ b/Makefile.am
@@ -1,7 +1,6 @@
+SUBDIRS=src Demos UnitTests
 if CONDITIONAL_BUILD_DEMOS
-SUBDIRS=src Extras Demos
-else
-SUBDIRS=src 
+SUBDIRS+=Extras
 endif
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = bullet.pc
diff --git a/UnitTests/BulletUnitTests/CMakeLists.txt b/UnitTests/BulletUnitTests/CMakeLists.txt
--- a/UnitTests/BulletUnitTests/CMakeLists.txt
+++ b/UnitTests/BulletUnitTests/CMakeLists.txt
@@ -9,7 +9,7 @@ INCLUDE_DIRECTORIES(
 
 LINK_LIBRARIES(
 	cppunit 
-	BulletMultiThreaded 
+#	BulletMultiThreaded 
 	BulletSoftBody
 	BulletDynamics  
 	BulletCollision 
diff --git a/UnitTests/BulletUnitTests/Main.cpp b/UnitTests/BulletUnitTests/Main.cpp
--- a/UnitTests/BulletUnitTests/Main.cpp
+++ b/UnitTests/BulletUnitTests/Main.cpp
@@ -40,7 +40,7 @@ int main(int argc, char* argv[])
   CPPUNIT_NS::CompilerOutputter outputter( &result, CPPUNIT_NS::stdCOut() );
   outputter.write(); 
 
-  getchar();
+  // getchar();
 
   return result.wasSuccessful() ? 0 : 1;
 }
diff --git a/UnitTests/BulletUnitTests/Makefile.am b/UnitTests/BulletUnitTests/Makefile.am
new file mode 100644
--- /dev/null
+++ b/UnitTests/BulletUnitTests/Makefile.am
@@ -0,0 +1,11 @@
+noinst_PROGRAMS=AppBulletUnitTests
+
+AppBulletUnitTests_SOURCES= Main.cpp TestBulletOnly.h TestLinearMath.h \
+	TestCholeskyDecomposition.cpp TestCholeskyDecomposition.h \
+	TestPolarDecomposition.cpp TestPolarDecomposition.h \
+	btCholeskyDecomposition.cpp btCholeskyDecomposition.h
+AppBulletUnitTests_CXXFLAGS=-I@top_srcdir@/src \
+	-I@top_srcdir@/UnitTests/cppunit/include $(CXXFLAGS)
+AppBulletUnitTests_LDADD=-L@top_builddir@/src \
+	-L@top_builddir@/UnitTests/cppunit -lcppunit -lBulletSoftBody \
+	-lBulletDynamics -lBulletCollision -lLinearMath
diff --git a/UnitTests/Makefile.am b/UnitTests/Makefile.am
new file mode 100644
--- /dev/null
+++ b/UnitTests/Makefile.am
@@ -0,0 +1 @@
+SUBDIRS=cppunit BulletUnitTests
diff --git a/UnitTests/cppunit/CMakeLists.txt b/UnitTests/cppunit/CMakeLists.txt
--- a/UnitTests/cppunit/CMakeLists.txt
+++ b/UnitTests/cppunit/CMakeLists.txt
@@ -48,15 +48,15 @@ ADD_LIBRARY(cppunit
 	src/cppunit/XmlOutputterHook.cpp
 	
 #plugin	
-	src/cppunit/BeOsDynamicLibraryManager.cpp
-	src/cppunit/DynamicLibraryManager.cpp
-	src/cppunit/DynamicLibraryManagerException.cpp
-	src/cppunit/PlugInManager.cpp
-	src/cppunit/PlugInParameters.cpp
-	src/cppunit/ShlDynamicLibraryManager.cpp
-	src/cppunit/TestPlugInDefaultImpl.cpp
-	src/cppunit/UnixDynamicLibraryManager.cpp
-	src/cppunit/Win32DynamicLibraryManager.cpp
+#	src/cppunit/BeOsDynamicLibraryManager.cpp
+#	src/cppunit/DynamicLibraryManager.cpp
+#	src/cppunit/DynamicLibraryManagerException.cpp
+#	src/cppunit/PlugInManager.cpp
+#	src/cppunit/PlugInParameters.cpp
+#	src/cppunit/ShlDynamicLibraryManager.cpp
+#	src/cppunit/TestPlugInDefaultImpl.cpp
+#	src/cppunit/UnixDynamicLibraryManager.cpp
+#	src/cppunit/Win32DynamicLibraryManager.cpp
 	
 #protector
 	src/cppunit/DefaultProtector.cpp
@@ -71,4 +71,4 @@ ADD_LIBRARY(cppunit
 	src/cppunit/XmlDocument.cpp
 	src/cppunit/XmlElement.cpp
 	
-)
\ No newline at end of file
+)
diff --git a/UnitTests/cppunit/Makefile.am b/UnitTests/cppunit/Makefile.am
new file mode 100644
--- /dev/null
+++ b/UnitTests/cppunit/Makefile.am
@@ -0,0 +1,45 @@
+noinst_LIBRARIES	= libcppunit.a
+
+libcppunit_a_CXXFLAGS = ${CXXFLAGS} -I@top_srcdir@/UnitTests/cppunit/include
+
+libcppunit_a_SOURCES =\
+	src/cppunit/AdditionalMessage.cpp \
+	src/cppunit/Asserter.cpp \
+	src/cppunit/Exception.cpp \
+	src/cppunit/Message.cpp \
+	src/cppunit/SourceLine.cpp \
+	src/cppunit/SynchronizedObject.cpp \
+	src/cppunit/Test.cpp \
+	src/cppunit/TestAssert.cpp \
+	src/cppunit/TestCase.cpp \
+	src/cppunit/TestComposite.cpp \
+	src/cppunit/TestFailure.cpp \
+	src/cppunit/TestLeaf.cpp \
+	src/cppunit/TestPath.cpp \
+	src/cppunit/TestResult.cpp \
+	src/cppunit/TestRunner.cpp \
+	src/cppunit/TestSuite.cpp \
+	src/cppunit/RepeatedTest.cpp \
+	src/cppunit/TestCaseDecorator.cpp \
+	src/cppunit/TestDecorator.cpp \
+	src/cppunit/TestSetUp.cpp \
+	src/cppunit/TestFactoryRegistry.cpp \
+	src/cppunit/TestNamer.cpp \
+	src/cppunit/TestSuiteBuilderContext.cpp \
+	src/cppunit/TypeInfoHelper.cpp \
+	src/cppunit/BriefTestProgressListener.cpp \
+	src/cppunit/TestResultCollector.cpp \
+	src/cppunit/TestSuccessListener.cpp \
+	src/cppunit/TextTestProgressListener.cpp \
+	src/cppunit/TextTestResult.cpp \
+	src/cppunit/CompilerOutputter.cpp \
+	src/cppunit/TextOutputter.cpp \
+	src/cppunit/XmlOutputter.cpp \
+	src/cppunit/XmlOutputterHook.cpp \
+	src/cppunit/DefaultProtector.cpp \
+	src/cppunit/Protector.cpp \
+	src/cppunit/ProtectorChain.cpp \
+	src/cppunit/TextTestRunner.cpp \
+	src/cppunit/StringTools.cpp \
+	src/cppunit/XmlDocument.cpp \
+	src/cppunit/XmlElement.cpp
diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -58,6 +58,16 @@ case "$host" in
                 ARCH_SPECIFIC_CFLAGS=""
                 ARCH_STRING="PowerPC"
                 ;;
+        arm-*)
+                AC_DEFINE(ARCH_ARM, 1, [Architecture is ARM])
+                ARCH_SPECIFIC_CFLAGS=""
+                ARCH_STIRNG="ARM"
+                ;;
+        le32-unknown-nacl*)
+                AC_DEFINE(ARCH_PNACL, 1, [Architecture is PNaCl])
+                ARCH_SPECIFIC_CFLAGS=""
+                ARCH_STIRNG="PNaCl"
+                ;;
         *)
                 AC_MSG_ERROR([Unknown Architecture])
                 ;;
@@ -163,7 +173,7 @@ CXXFLAGS="$ARCH_SPECIFIC_CFLAGS $CXXFLAGS $CFLAGS"
 #----------------------------------------------------------------------------
 # Emit generated files.
 #----------------------------------------------------------------------------
-AC_CONFIG_FILES([bullet.pc Makefile Demos/Makefile Demos/SoftDemo/Makefile Demos/AllBulletDemos/Makefile Demos/MultiThreadedDemo/Makefile  Demos/OpenGL/Makefile Demos/BasicDemo/Makefile Demos/CcdPhysicsDemo/Makefile Demos/VehicleDemo/Makefile Demos/TerrainDemo/Makefile src/Makefile Extras/Makefile])
+AC_CONFIG_FILES([bullet.pc Makefile Demos/Makefile Demos/SoftDemo/Makefile Demos/AllBulletDemos/Makefile Demos/MultiThreadedDemo/Makefile  Demos/OpenGL/Makefile Demos/BasicDemo/Makefile Demos/CcdPhysicsDemo/Makefile Demos/VehicleDemo/Makefile Demos/TerrainDemo/Makefile Demos/HelloWorld/Makefile src/Makefile Extras/Makefile UnitTests/Makefile UnitTests/cppunit/Makefile UnitTests/BulletUnitTests/Makefile])
 AC_OUTPUT
 
 AC_MSG_NOTICE([
