diff --git a/configure b/configure
--- a/configure
+++ b/configure
@@ -11863,7 +11863,7 @@ else
         case $host in
         *)
                 DEBUG="-g -Wall -Wno-parentheses -DDEBUG -D__NO_MATH_INLINES"
-                CFLAGS="-Wall -Wno-parentheses -O3 -fforce-addr -fomit-frame-pointer -finline-functions -funroll-loops"
+                CFLAGS="-Wall -Wno-parentheses -O3 -fomit-frame-pointer -finline-functions -funroll-loops"
                 PROFILE="-Wall -Wno-parentheses -pg -g -O3 -fno-inline-functions -DDEBUG";;
         esac
 fi
diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -168,7 +168,7 @@ else
         case $host in 
         *)
                 DEBUG="-g -Wall -Wno-parentheses -DDEBUG -D__NO_MATH_INLINES"
-                CFLAGS="-Wall -Wno-parentheses -O3 -fforce-addr -fomit-frame-pointer -finline-functions -funroll-loops"
+                CFLAGS="-Wall -Wno-parentheses -O3 -fomit-frame-pointer -finline-functions -funroll-loops"
                 PROFILE="-Wall -Wno-parentheses -pg -g -O3 -fno-inline-functions -DDEBUG";;
         esac
 fi
diff --git a/lib/cpu.c b/lib/cpu.c
--- a/lib/cpu.c
+++ b/lib/cpu.c
@@ -20,6 +20,7 @@
 
 #include "cpu.h"
 
+
 #if !defined(OC_X86_ASM)
 static ogg_uint32_t oc_cpu_flags_get(void){
   return 0;
@@ -130,6 +131,7 @@ static ogg_uint32_t oc_cpu_flags_get(void){
   ogg_uint32_t ebx;
   ogg_uint32_t ecx;
   ogg_uint32_t edx;
+# if !defined(__native_client__)
 # if !defined(__amd64__)&&!defined(__x86_64__)
   /*Not all x86-32 chips support cpuid, so we have to check.*/
 #  if !defined(_MSC_VER)
@@ -154,6 +156,7 @@ static ogg_uint32_t oc_cpu_flags_get(void){
   /*No cpuid.*/
   if(eax==ebx)return 0;
 # endif
+# endif
   cpuid(0,eax,ebx,ecx,edx);
   /*         l e t n          I e n i          u n e G*/
   if(ecx==0x6C65746E&&edx==0x49656E69&&ebx==0x756E6547||
