diff --git a/Makefile b/Makefile
--- a/Makefile
+++ b/Makefile
@@ -3,7 +3,12 @@ all:	lib
 src/libagg.a:
 	cd src; make
 
-lib:	src/libagg.a
+font_freetype/libaggfontfreetype.a:
+	cd font_freetype; make
+
+lib:	src/libagg.a font_freetype/libaggfontfreetype.a
 
 clean:
 	cd src; make clean
+	cd font_freetype; make clean
+
diff --git a/Makefile.in.NativeClient b/Makefile.in.NativeClient
new file mode 100644
--- /dev/null
+++ b/Makefile.in.NativeClient
@@ -0,0 +1,7 @@
+AGGLIBS= -lagg
+CXX = $(NACLCXX)
+C = $(NACLCC)
+LIB = $(NACLAR) cr
+
+.PHONY : clean
+
diff --git a/font_freetype/Makefile b/font_freetype/Makefile
new file mode 100644
--- /dev/null
+++ b/font_freetype/Makefile
@@ -0,0 +1,21 @@
+include ../Makefile.in.NativeClient
+
+CXXFLAGS= $(AGGCXXFLAGS) -I../include -L./ \
+  -DFT_FREETYPE_H="<freetype/freetype.h>"
+
+SRC_CXX=\
+agg_font_freetype.cpp 
+
+OBJ=$(SRC_CXX:.cpp=.o)
+
+all:	$(OBJ)
+	$(LIB) libaggfontfreetype.a $(OBJ) 
+
+clean:
+	rm -f *.o *.a
+
+%.o: %.cpp
+	$(CXX) -c $(CXXFLAGS) $*.cpp -o $@
+
+%.o: %.c
+	$(C) -c $(CXXFLAGS) $*.c -o $@
diff --git a/include/agg_renderer_outline_aa.h b/include/agg_renderer_outline_aa.h
--- a/include/agg_renderer_outline_aa.h
+++ b/include/agg_renderer_outline_aa.h
@@ -1375,7 +1375,6 @@ namespace agg
         //---------------------------------------------------------------------
         void profile(const line_profile_aa& prof) { m_profile = &prof; }
         const line_profile_aa& profile() const { return *m_profile; }
-        line_profile_aa& profile() { return *m_profile; }
 
         //---------------------------------------------------------------------
         int subpixel_width() const { return m_profile->subpixel_width(); }
diff --git a/src/Makefile b/src/Makefile
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,4 +1,4 @@
-include ../Makefile.in.$(shell uname)
+include ../Makefile.in.NativeClient
 
 CXXFLAGS= $(AGGCXXFLAGS) -I../include -L./
 
