diff --git a/configure b/configure
--- a/configure
+++ b/configure
@@ -18510,7 +18523,11 @@ ac_config_commands="$ac_config_commands libtool_hack"
 # OSX/Win32 rules are different.
 platform_win32=no
 platform_darwin=no
-LIBEXT=so
+if [ "$enable_shared" = "no" ]; then
+  LIBEXT=a
+else
+  LIBEXT=so
+fi
 case $host_os in
     cygwin*|mingw*)
 	LIBEXT=dll.a
diff --git a/src/Makefile.in b/src/Makefile.in
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -1438,7 +1438,7 @@ uninstall-am: uninstall-libLTLIBRARIES uninstall-local
 @BUILD_XAW6_TRUE@@PLATFORM_DARWIN_TRUE@@PLATFORM_WIN32_FALSE@	-rm -f $(DESTDIR)$(libdir)/libXaw.6.@LIBEXT@
 @BUILD_XAW6_TRUE@@PLATFORM_DARWIN_TRUE@@PLATFORM_WIN32_FALSE@	(cd $(DESTDIR)$(libdir) && ln -s libXaw6.6.@LIBEXT@ libXaw.6.@LIBEXT@)
 @BUILD_XAW6_TRUE@@PLATFORM_DARWIN_FALSE@@PLATFORM_WIN32_FALSE@	-rm -f $(DESTDIR)$(libdir)/libXaw.@LIBEXT@.6
-@BUILD_XAW6_TRUE@@PLATFORM_DARWIN_FALSE@@PLATFORM_WIN32_FALSE@	(cd $(DESTDIR)$(libdir) && ln -s libXaw6.@LIBEXT@.6 libXaw.@LIBEXT@.6)
+@BUILD_XAW6_TRUE@@PLATFORM_DARWIN_FALSE@@PLATFORM_WIN32_FALSE@	(cd $(DESTDIR)$(libdir) && if [ -f libXaw6.@LIBEXT@.6 ]; then ln -s libXaw6.@LIBEXT@.6 libXaw.@LIBEXT@.6; fi)
 
 @BUILD_XAW6_TRUE@@PLATFORM_WIN32_FALSE@uninstall-local::
 @BUILD_XAW6_TRUE@@PLATFORM_DARWIN_TRUE@@PLATFORM_WIN32_FALSE@	-rm -f $(DESTDIR)$(libdir)/libXaw.6.@LIBEXT@
@@ -1448,7 +1448,7 @@ uninstall-am: uninstall-libLTLIBRARIES uninstall-local
 @BUILD_XAW7_TRUE@@PLATFORM_DARWIN_TRUE@@PLATFORM_WIN32_FALSE@	-rm -f $(DESTDIR)$(libdir)/libXaw.7.@LIBEXT@
 @BUILD_XAW7_TRUE@@PLATFORM_DARWIN_TRUE@@PLATFORM_WIN32_FALSE@	(cd $(DESTDIR)$(libdir) && ln -s libXaw7.7.@LIBEXT@ libXaw.7.@LIBEXT@)
 @BUILD_XAW7_TRUE@@PLATFORM_DARWIN_FALSE@@PLATFORM_WIN32_FALSE@	-rm -f $(DESTDIR)$(libdir)/libXaw.@LIBEXT@.7
-@BUILD_XAW7_TRUE@@PLATFORM_DARWIN_FALSE@@PLATFORM_WIN32_FALSE@	(cd $(DESTDIR)$(libdir) && ln -s libXaw7.@LIBEXT@.7 libXaw.@LIBEXT@.7)
+@BUILD_XAW7_TRUE@@PLATFORM_DARWIN_FALSE@@PLATFORM_WIN32_FALSE@	(cd $(DESTDIR)$(libdir) && if [ -f libXaw7.@LIBEXT@.7 ]; then ln -s libXaw7.@LIBEXT@.7 libXaw.@LIBEXT@.7; fi)
 
 @BUILD_XAW7_TRUE@@PLATFORM_WIN32_FALSE@uninstall-local::
 @BUILD_XAW7_TRUE@@PLATFORM_DARWIN_TRUE@@PLATFORM_WIN32_FALSE@	-rm -f $(DESTDIR)$(libdir)/libXaw.7.@LIBEXT@
diff --git a/src/TextAction.c b/src/TextAction.c
--- a/src/TextAction.c
+++ b/src/TextAction.c
@@ -47,6 +47,10 @@ in this Software without prior written authorization from The Open Group.
 #include "Private.h"
 #include "XawI18n.h"
 
+#if defined(__native_client__)
+# include <sys/select.h>
+#endif
+
 #define SrcScan			XawTextSourceScan
 #define FindDist		XawTextSinkFindDistance
 #define FindPos			XawTextSinkFindPosition
