diff --git a/common/cpu.c b/common/cpu.c
--- a/common/cpu.c
+++ b/common/cpu.c
@@ -93,7 +93,7 @@ const x264_cpu_name_t x264_cpu_names[] =
     {"", 0},
 };
 
-#if (ARCH_PPC && SYS_LINUX) || (ARCH_ARM && !HAVE_NEON)
+#if (ARCH_PPC && SYS_LINUX) || (ARCH_ARM && HAVE_ARMV6 && !HAVE_NEON)
 #include <signal.h>
 #include <setjmp.h>
 static sigjmp_buf jmpbuf;
diff --git a/configure b/configure
--- a/configure
+++ b/configure
@@ -547,6 +547,16 @@ case $host_os in
         LDFLAGSCLI="$LDFLAGSCLI -lshell32"
         [ $compiler = GNU ] && RC="${RC-${cross_prefix}windres}" || RC="${RC-rc}"
         ;;
+    *nacl*)
+        SYS="NACL"
+        define HAVE_MALLOC_H
+        if [[ $(basename ${CC}) =~ pnacl-* ]]; then
+            EXE=.pexe
+        else
+            EXE=.nexe
+        fi
+        libm="-lm"
+        ;;
     sunos*|solaris*)
         SYS="SunOS"
         define HAVE_MALLOC_H
@@ -773,11 +783,14 @@ ASFLAGS="$ASFLAGS -DSTACK_ALIGNMENT=$stack_alignment"
 if [ $compiler = GNU ]; then
     echo "int i[2] = {0x42494745,0}; double f[2] = {0x1.0656e6469616ep+102,0};" > conftest.c
     $CC $CFLAGS conftest.c -c -o conftest.o 2>/dev/null || die "endian test failed"
+    # there is no pnacl-strings binary.
+    if [[ ! $(basename ${CC}) =~ pnacl-* ]]; then
     if (${cross_prefix}strings -a conftest.o | grep -q BIGE) && (${cross_prefix}strings -a conftest.o | grep -q FPendian) ; then
         define WORDS_BIGENDIAN
     elif !(${cross_prefix}strings -a conftest.o | grep -q EGIB && ${cross_prefix}strings -a conftest.o | grep -q naidnePF) ; then
         die "endian test failed"
     fi
+    fi
 fi
 
 # autodetect options that weren't forced nor disabled
