diff --git a/lib/hostip4.c b/lib/hostip4.c
--- a/lib/hostip4.c
+++ b/lib/hostip4.c
@@ -115,6 +115,12 @@ Curl_addrinfo *Curl_getaddrinfo(struct connectdata *conn,
 
 #if defined(CURLRES_IPV4) && !defined(CURLRES_ARES)
 
+#ifdef __native_client__
+// nacl_io does not yet support gethostbyname_r().
+// TODO(sbc): remove this once it is added to nacl_io: http://crbug.com/387474
+#undef HAVE_GETHOSTBYNAME_R
+#endif
+
 /*
  * Curl_ipv4_resolve_r() - ipv4 threadsafe resolver function.
  *
diff --git a/src/Makefile.in b/src/Makefile.in
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -749,7 +749,7 @@ clean-binPROGRAMS:
 
 curl$(EXEEXT): $(curl_OBJECTS) $(curl_DEPENDENCIES) $(EXTRA_curl_DEPENDENCIES) 
 	@rm -f curl$(EXEEXT)
-	$(AM_V_CCLD)$(curl_LINK) $(curl_OBJECTS) $(curl_LDADD) $(LIBS)
+	$(AM_V_CCLD)$(curl_LINK) $(curl_OBJECTS) $(curl_LDADD) $(LIBS) $(EXTRA_LIBS)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
diff --git a/src/tool_main.c b/src/tool_main.c
--- a/src/tool_main.c
+++ b/src/tool_main.c
@@ -213,6 +213,10 @@ static void main_free(struct GlobalConfig *config)
   config->last = NULL;
 }
 
+#ifdef __native_client__
+#define main nacl_main
+#endif
+
 /*
 ** curl tool main function.
 */
