diff --git a/rldefs.h b/rldefs.h
--- a/rldefs.h
+++ b/rldefs.h
@@ -34,6 +34,22 @@
 #  undef HAVE_STRCOLL
 #endif
 
+#if defined (__native_client__) && !defined (_POSIX_VERSION)
+/* Currently the newlib toolchain in NaCl doesn't define _POSIX_VERSION
+ * which causes readline to fall back to using ioctl() rather than
+ * tcgetattr().  TODO(sbc): remove this patch once either ioctl() is
+ * supported on tty nodes or _POSIX_VERSION is defined by newlib headers. */
+#define _POSIX_VERSION 199309L
+#endif
+
+#if defined (__native_client__) && !defined (__GLIBC__) && !defined TCOON
+// TODO(sbc): remove once these are added to termios.h in NaCl
+#define TCOOFF    0
+#define TCOON     1
+#define TCIOFF    2
+#define TCION     3
+#endif
+
 #if defined (_POSIX_VERSION) && !defined (TERMIOS_MISSING)
 #  define TERMIOS_TTY_DRIVER
 #else
