diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -19,10 +19,10 @@ set(BOX2D_VERSION 2.1.0)
 # The Box2D library.
 add_subdirectory(Box2D)
 
-if(BOX2D_BUILD_EXAMPLES)
-  # HelloWorld console example.
-  add_subdirectory(HelloWorld)
+# HelloWorld console example.
+add_subdirectory(HelloWorld)
 
+if(BOX2D_BUILD_EXAMPLES)
   # Testbed and dependencies.
   find_package(OpenGL REQUIRED)
   add_subdirectory(freeglut)
@@ -32,4 +32,4 @@ endif(BOX2D_BUILD_EXAMPLES)
 
 if(BOX2D_INSTALL_DOC)
   install(DIRECTORY Documentation DESTINATION share/doc/Box2D PATTERN ".svn" EXCLUDE)
-endif(BOX2D_INSTALL_DOC)
\ No newline at end of file
+endif(BOX2D_INSTALL_DOC)
