diff --git a/Makefile.in b/Makefile.in
--- a/Makefile.in
+++ b/Makefile.in
@@ -103,6 +103,7 @@ mkdir-init:
 CC = @CC@
 CXX = @CXX@
 LIBTOOL = @LIBTOOL@
+LIBS = @LIBS@
 
 INCLUDES = -I$(srcdir)/lib -I.
 LDFLAGS = @LDFLAGS@
@@ -143,7 +144,7 @@ xmlwf/xmlfile.o: xmlwf/xmlfile.c
 xmlwf/codepage.o: xmlwf/codepage.c
 xmlwf/@FILEMAP@.o: xmlwf/@FILEMAP@.c
 xmlwf/xmlwf: $(XMLWF_OBJS) $(LIBRARY)
-	$(LINK_EXE) $(XMLWF_OBJS) $(LIBRARY)
+	$(LINK_EXE) $(XMLWF_OBJS) $(LIBRARY) $(LIBS)
 
 examples/elements.o: examples/elements.c
 examples/elements: examples/elements.o $(LIBRARY)
