diff --git a/source/configure b/source/configure
--- a/source/configure
+++ b/source/configure
@@ -4938,6 +4938,7 @@ rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
 *-apple-darwin*)	icu_cv_host_frag=mh-darwin ;;
 *-*-beos)       icu_cv_host_frag=mh-beos ;;
 *-*-haiku)      icu_cv_host_frag=mh-haiku ;;
+*-*-nacl)      icu_cv_host_frag=mh-linux ;;
 *-*-irix*)	icu_cv_host_frag=mh-irix ;;
 *-dec-osf*) icu_cv_host_frag=mh-alpha-osf ;;
 *-*-nto*)	icu_cv_host_frag=mh-qnx ;;
@@ -7406,6 +7407,7 @@ ICUDATA_CHAR=$U_ENDIAN_CHAR
 case "${host}" in
 	*-*-solaris*) 	platform=U_SOLARIS ;;
 	*-*-linux*|*-*-gnu|*-*-k*bsd*-gnu|*-*-kopensolaris*-gnu)        platform=U_LINUX ;;
+	*-*-nacl)        platform=U_NACL ;;
 	*-*-*bsd*|*-*-dragonfly*) 	platform=U_BSD ;;
 	*-*-aix*) 	platform=U_AIX ;;
 	*-*-hpux*) 	platform=U_HPUX ;;
diff --git a/source/tools/gendict/gendict.cpp b/source/tools/gendict/gendict.cpp
--- a/source/tools/gendict/gendict.cpp
+++ b/source/tools/gendict/gendict.cpp
@@ -43,7 +43,7 @@ static int elapsedTime() {
   return (int)uprv_floor((uprv_getRawUTCtime()-startTime)/1000.0);
 }
 
-#if U_PLATFORM_IMPLEMENTS_POSIX && !U_PLATFORM_HAS_WIN32_API
+#if U_PLATFORM_IMPLEMENTS_POSIX && !U_PLATFORM_HAS_WIN32_API && !defined __native_client__
 
 #include <signal.h>
 #include <unistd.h>
diff --git a/source/tools/toolutil/pkg_genc.c b/source/tools/toolutil/pkg_genc.c
--- a/source/tools/toolutil/pkg_genc.c
+++ b/source/tools/toolutil/pkg_genc.c
@@ -672,9 +672,14 @@ getArchitecture(uint16_t *pCPU, uint16_t *pBits, UBool *pIsBigEndian, const char
         /* set defaults */
 #ifdef U_ELF
         /* set EM_386 because elf.h does not provide better defaults */
+#if defined(__x86_64__)
+        *pCPU=EM_X86_64;
+#else
         *pCPU=EM_386;
-        *pBits=32;
-        *pIsBigEndian=(UBool)(U_IS_BIG_ENDIAN ? ELFDATA2MSB : ELFDATA2LSB);
+#endif
+        *pBits= *pCPU==EM_386 ? 32 : 64;
+        *pIsBigEndian=(UBool)(U_IS_BIG_ENDIAN ? TRUE : FALSE);
+        printf("U_IS_BIG_ENDIAN : %d\n", U_IS_BIG_ENDIAN);
 #elif U_PLATFORM_HAS_WIN32_API
 /* _M_IA64 should be defined in windows.h */
 #   if defined(_M_IA64)
@@ -1022,7 +1027,7 @@ writeObjectCode(const char *filename, const char *destdir, const char *optEntryP
 
     /* deal with options, files and the entry point name */
     getArchitecture(&cpu, &bits, &makeBigEndian, optMatchArch);
-    printf("genccode: --match-arch cpu=%hu bits=%hu big-endian=%d\n", cpu, bits, makeBigEndian);
+    printf("genccode: --match-arch=%s cpu=%hu bits=%hu big-endian=%d\n", optMatchArch, cpu, bits, makeBigEndian);
 #if U_PLATFORM_HAS_WIN32_API
     if(cpu==IMAGE_FILE_MACHINE_I386) {
         entryOffset=1;
