diff --git a/makefile.unix b/makefile.unix
--- a/makefile.unix
+++ b/makefile.unix
@@ -3,16 +3,20 @@
 # Tom St Denis
 
 # Compiler and Linker Names
-CC=cc
-LD=ld
+CC?=cc
+LD?=ld
+RANLIB?=ranlib
 
 # Archiver [makes .a files]
-AR=ar
-ARFLAGS=r
+AR?=ar
+ARFLAGS?=r
 
 # Compilation flags. Note the += does not write over the user's CFLAGS!
 CFLAGS = -c -I./testprof/ -I./src/headers/ -DLTC_SOURCE -O2 ${CFLAGS_OPTS} -o $@
 
+CFLAGS += -DGMP_DESC -DUSE_GMP -DLTC_NO_ASM
+EXTRALIBS ?= -lgmp
+
 LIBNAME=libtomcrypt.a
 LIBTEST=libtomcrypt_prof.a
 LIBTEST_S=$(LIBTEST)
diff --git a/src/headers/tomcrypt_cfg.h b/src/headers/tomcrypt_cfg.h
--- a/src/headers/tomcrypt_cfg.h
+++ b/src/headers/tomcrypt_cfg.h
@@ -70,7 +70,7 @@ LTC_EXPORT int   LTC_CALL XSTRCMP(const char *s1, const char *s2);
 #endif
 
 /* detect amd64 */
-#if !defined(__STRICT_ANSI__) && defined(__x86_64__)
+#if !defined(__STRICT_ANSI__) && defined(__x86_64__) && !defined(__native_client__)
    #define ENDIAN_LITTLE
    #define ENDIAN_64BITWORD
    #define LTC_FAST
diff --git a/src/headers/tomcrypt_macros.h b/src/headers/tomcrypt_macros.h
--- a/src/headers/tomcrypt_macros.h
+++ b/src/headers/tomcrypt_macros.h
@@ -7,10 +7,10 @@
    typedef unsigned long long ulong64;
 #endif
 
-/* this is the "32-bit at least" data type 
- * Re-define it to suit your platform but it must be at least 32-bits 
+/* this is the "32-bit at least" data type
+ * Re-define it to suit your platform but it must be at least 32-bits
  */
-#if defined(__x86_64__) || (defined(__sparc__) && defined(__arch64__))
+#if !defined(__native_client__) && defined(__x86_64__) || (defined(__sparc__) && defined(__arch64__))
    typedef unsigned ulong32;
 #else
    typedef unsigned long ulong32;
@@ -67,7 +67,7 @@
 
 #ifdef ENDIAN_LITTLE
 
-#if !defined(LTC_NO_BSWAP) && (defined(INTEL_CC) || (defined(__GNUC__) && (defined(__DJGPP__) || defined(__CYGWIN__) || defined(__MINGW32__) || defined(__i386__) || defined(__x86_64__))))
+#if !defined(LTC_NO_BSWAP) && (defined(INTEL_CC) || (defined(__GNUC__) && (defined(__DJGPP__) || defined(__CYGWIN__) || defined(__MINGW32__) || defined(__i386__) || defined(__x86_64__) && !defined(__native_client__))))
 
 #define STORE32H(x, y)           \
 asm __volatile__ (               \
@@ -98,7 +98,7 @@ asm __volatile__ (             \
 
 
 /* x86_64 processor */
-#if !defined(LTC_NO_BSWAP) && (defined(__GNUC__) && defined(__x86_64__))
+#if !defined(LTC_NO_BSWAP) && (defined(__GNUC__) && defined(__x86_64__) && !defined(__native_client__))
 
 #define STORE64H(x, y)           \
 asm __volatile__ (               \
@@ -341,7 +341,7 @@ static inline unsigned RORc(unsigned word, const int i)
 
 
 /* 64-bit Rotates */
-#if !defined(__STRICT_ANSI__) && defined(__GNUC__) && defined(__x86_64__) && !defined(LTC_NO_ASM)
+#if !defined(__STRICT_ANSI__) && defined(__GNUC__) && defined(__x86_64__) && !defined(LTC_NO_ASM) && !defined(__native_client__)
 
 static inline unsigned long ROL64(unsigned long word, int i)
 {
