diff --git a/src/tools/oggz-chop/httpdate.c b/src/tools/oggz-chop/httpdate.c
--- a/src/tools/oggz-chop/httpdate.c
+++ b/src/tools/oggz-chop/httpdate.c
@@ -5,6 +5,10 @@
 #include <string.h>
 #include <time.h>
 
+#ifdef _NEWLIB_VERSION
+#define timezone _timezone
+#endif
+
 #define HTTPDATE_FMT "%3s, %02d %s %4d %02d:%02d:%02d GMT"
 
 static char * wdays[] = {"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};
diff --git a/src/tools/oggz-chop/httpdate_test.c b/src/tools/oggz-chop/httpdate_test.c
--- a/src/tools/oggz-chop/httpdate_test.c
+++ b/src/tools/oggz-chop/httpdate_test.c
@@ -7,6 +7,10 @@
 
 #include "httpdate.h"
 
+#ifdef _NEWLIB_VERSION
+#define timezone _timezone
+#endif
+
 int
 main (int argc, char * argv[])
 {
