diff --git a/Xfuncs.h b/Xfuncs.h
--- a/Xfuncs.h
+++ b/Xfuncs.h
@@ -44,7 +44,7 @@ void bcopy();
 #    define bcmp(b1,b2,len) memcmp(b1, b2, len)
 #   else
 #    include <string.h>
-#    if defined(__SCO__) || defined(__sun) || defined(__UNIXWARE__) || defined(__CYGWIN__)
+#    if defined(__SCO__) || defined(__sun) || defined(__UNIXWARE__) || defined(__CYGWIN__) || defined(__native_client__)
 #     include <strings.h>
 #    endif
 #    define _XFUNCS_H_INCLUDED_STRING_H
diff --git a/Xos.h b/Xos.h
--- a/Xos.h
+++ b/Xos.h
@@ -60,7 +60,7 @@ in this Software without prior written authorization from The Open Group.
  */
 
 # include <string.h>
-# if defined(__SCO__) || defined(__UNIXWARE__) || defined(__sun) || defined(__CYGWIN__)
+# if defined(__SCO__) || defined(__UNIXWARE__) || defined(__sun) || defined(__CYGWIN__) || defined(__native_client__)
 #  include <strings.h>
 # else
 #  ifndef index
